eos-update-flatpak-repos

```diff
@@ -1,10 +1,379 @@
+
+    # Write the new GVariant to a temporary file and replace the original one.
+    with tempfile.NamedTemporaryFile() as tmp_deploy_file:
+        logging.info("Building new deploy file for {}, setting [{}] to: {}"
+                     .format(deploy_file_path, idx, new_val))
+
+        # Write temporary file to disk and overwrite the original one with it.
+        dest_file = Gio.File.new_for_path(tmp_deploy_file.name)
+        out_stream = dest_file.append_to(Gio.FileCreateFlags.NONE, None)
+        out_stream.write_bytes(new_variant.get_data_as_bytes())
+        out_stream.close()
+
+        os.chmod(tmp_deploy_file.name, 0o644)
+        shutil.copy(tmp_deploy_file.name, deploy_file_path, follow_symlinks=True)
+
+FLATPAKS_TO_MIGRATE = [
+    # switch EknServices to eos-sdk remote, migrating from old eos-apps remote (T17863)
+    {
+        'name': 'com.endlessm.EknServices',
+        'kind': Flatpak.RefKind.APP,
+        'old-origin': 'eos-apps',
+        'new-origin': 'eos-sdk'
+    },
+    # fix up image-builder bug where com.endlessm.apps.* runtimes were
+    # mistakenly installed with eos-runtimes origin (T18366)
+    {
+        'name': 'com.endlessm.apps.Platform',
+        'kind': Flatpak.RefKind.RUNTIME,
+        'old-origin': 'eos-runtimes',
+        'new-origin': 'eos-sdk'
+    },
+    {
+        'name': 'com.endlessm.apps.Sdk',
+        'kind': Flatpak.RefKind.RUNTIME,
+        'old-origin': 'eos-runtimes',
+        'new-origin': 'eos-sdk'
+    },
+    # migrate Steam from eos-apps to flathub (T20322)
+    {
+        'name': 'com.valvesoftware.Steam',
+        'kind': Flatpak.RefKind.APP,
+        'old-branch': 'eos3',
+        'new-branch': 'stable',
+        'old-origin': 'eos-apps',
+        'new-origin': 'flathub'
+    }
+]
+
+def _filter_refs(refs, name=None, prefix=None, kind=None, arch=None,
+                 branch=None, origin=None):
+    ret = []
+    for ref in refs:
+        if kind and kind != ref.get_kind():
+            continue
+        if name and name != ref.get_name():
+            continue
+        if prefix and not ref.get_name().startswith(prefix):
+            continue
+        if arch and arch != ref.get_arch():
+            continue
+        if branch and branch != ref.get_branch():
+            continue
+        if origin and origin != ref.get_origin():
+            continue
+        ret += [ref]
+    return ret
+
+def _migrate_installed_flatpaks():
+    insts = Flatpak.get_system_installations()
+    inst = insts[0]
+
+    refs = inst.list_installed_refs()
+    for migrate in FLATPAKS_TO_MIGRATE:
+        name = migrate['name']
+        kind = migrate['kind']
+        old_branch = migrate.get('old-branch')
+        old_origin = migrate.get('old-origin')
+
+        matching_refs = _filter_refs(refs, name=name, kind=kind,
+                                     branch=old_branch, origin=old_origin)
+        # also search for name.* runtimes to migrate to catch extensions
```

Comment:
Could you just do one filter here, with `prefix=name`?

Reply:
No, we need to look for apps and runtimes separately - and even in the runtime case, this would inadvertently include com.example.runtime2 when we only wanted com.example.runtime and com.example.runtime.extension.